### DISCUSSION TITLE
[Feature Request] PWA (Progressive web app) support

### DISCUSSION THREAD

**Discussion by edersong** [music-assistant/support#1788] [Feature requests and ideas]
It would be a nice feature to be able to install the website as an app on Android and iOS.

Ref.: https://blog.heroku.com/how-to-make-progressive-web-app

**Comment by OzGav** [comment]
There is s desktop app in Alpha. There is no developer capacity to develop apps for other platforms although it is on the road map. If you want to take on the job that would be great.

**Comment by marcelveldt** [comment]
The MA frontend is already a PWA so .... 

Navigate to the MA interface directly (bypass HA ingress) http://youripgoeshere:8095

But you may have to expose the webserver in the settings to allow this.

**Reply by jcbevns** [reply]
Just navigating to this is not a PWA. But as i add to Homescreen on Android, it still opens with firefox as a tab doesn't appear fullscreen like other PWA apps, so wondering if this is still supported.

**Reply by jcbevns** [reply]
Do we need `display: fullscreen` here or something?

https://vscode.dev/github/music-assistant/frontend/blob/main/vite.config.ts#L41

**Reply by marcelveldt** [reply]
You are the first to report that - it works fine for me on multiple devices.

**Reply by cephos** [reply]
It doesn't work on my android phone too. Just updated to version 2.4.0 as a ha-addon. Tried directly over http and with https through a nginx reverse proxy, both ways didnt work. Will try to reproduce it with an emulator when I find some time on the desktop.

**Reply by ItsHarper** [reply]
MDN says PWAs need to be hosted via HTTPS: https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Guides/Making_PWAs_installable#https_localhost_or_loopback_are_required, so we'd expect this to only work with a reverse proxy. I haven't set that up, so it makes sense that it's not working for me, and maybe @jcbevns. Doesn't explain why that didn't work for @cephos though.

**Reply by cephos** [reply]
Just an idea: I configured a basic-auth on the reverse proxy. Perhaps that has a negative inpact on the service workers that are needed for pwa functionality. Will try to disable it temporarly and see if it starts to work then and report back.. if thats the case I have to find another way to inplement authentication for an web-exposed domain (If someone has an idea how to do that...?)

**Reply by cephos** [reply]
Guess the service worker would need some modification for my use case (https://thatemil.com/blog/2018/02/21/pwa-basic-auth/). When I find time i'll create a pull request so its fixed for all who have the same setup as I have.
Background of my setup is that I want to have a direct pwa-app to musicassistant on our android phones, so we ddont have to navigate in ha to the ma-dsshboard-menu-entry each. time to switch music.

**Reply by ItsHarper** [reply]
Wow I hadn't even processed that Music Assistant doesn't have an authentication system

**Reply by cephos** [reply]
After some research I found we probably need the following option in the vitepwa config:

useCredentials: true

Tried now to find the config file in an attached console of my prod ma-installation in ha os to make quick test, but without succcess. Can someone point me to the right path? Or is this 'compiled' in and only available in a dev container?

**Reply by jace** [reply]
This isn't working for me either. I've setup a Cloudflare Tunnel for HTTPS and authentication. Confirmed from view-source that the manifest is still present. Chrome doesn't recognise it as a PWA, nor do other Android browsers.

**Reply by jace** [reply]
I've found a workaround for Cloudflare Tunnel. The trick is to make a separate Access Application pointing specifically to the manifest URL (`<domain>/manifest.webmanifest`) and link it to a new Bypass Policy that applies to Everyone. The manifest no longer requires authentication. Chrome Android showed an Install App prompt rightaway.

The downside is that the PWA isn't LAN-only.

**Reply by sameleff** [reply]
Were you able to get the PWA app working? I am using Firefox on iOS and reverse proxying the Music Assistant webserver with nginx. No auth is enabled in nginx. I have tried adding the shortcut by connecting directly to the ip:port and also to the proxied url - in both cases, the shortcut opens a new tab in Firefox rather than a full screen and standalone PWA app. Other services I self-host and configured analogously (e.g. OpenWebUI) successfully open in a standalone app when setup in the same manner through Firefox.

**Reply by drewsteinacher** [reply]
I'm also unable to get the PWA working with Firefox on Android and reverse proxying to the Music Assistant webserver with Caddy.

FWIW, these all seem related:
* https://github.com/music-assistant/support/issues/2906#issuecomment-2346075796
* https://github.com/orgs/music-assistant/discussions/3747

I'm not sure if the `useCredentials` thing works more generally than the situation described by cephos above.

But the Cloudlfare Tunnel thing won't work for me and my purely-local setup.

Lastly, I would like to echo that having a working PWA is much, much better than using the browser or waiting for a native app to be developed (but that seems pointless; surely it's easier to get the PWA install working than building a separate native app).

**Reply by drewsteinacher** [reply]
Although I tried poking at the useCredentials thing in a dev environment, ultimately, that's not what I needed to get my PWA installed.

I carefully followed all of the steps [in this guide](https://m0x2a.dreamymatrix.com/caddy-as-internal-ca-and-reverse-proxy/) and made sure to copy the CA root certificate from `/var/lib/caddy/.local/share/caddy/pki/authorities/local/root.crt` as listed on that guide, which was different from the location I was copying in my previous attempts (not sure why it was different -- might have been a mistake on my end in previous attempts. Be sure to use `shasum` or similar to make sure it's right!).

After installing [PWAs for Firefox](https://github.com/filips123/PWAsForFirefox), I could get the PWA to install on my desktop Ubuntu machine in Firefox, even without the root cert installed (via a security exception, of course).

On my Android device, I got it to work by copying/installing the root cert in the system settings and then enabling developer mode in Firefox for Android and then enabling "Use third party CA certificates", as described in that guide.

TL;DR: No code changes are _technically_ necessary to get the PWA for MA to install if you have a reverse proxy configured _properly_ and you properly and _completely_ configure your device/browser to trust the relevant root CA.

Hopefully this helps others who wish to have a dedicated "app" for MA on their devices while waiting for [the client app](https://github.com/music-assistant/kmp-client-app) to be ready for general use :)
